### ISSUE TITLE
Merging runs question

### ISSUE THREAD

**Issue by drinkermoth** [benjjneb/dada2#1103]
Hello, I have sequences around 800 samples for ITS1 and 16S on an Illumina Miseq and Illumina HiSeq run. I have analysed them using the same dada2 filt and trim methods and have produced seqtab and taxa files for each locus and run. 

I've merged my fungal ASVs but found that this has only combined the two seqtab tables in such a way that exactly matching ASVs are pooled but non-exact matches by rowname are not. I thought that pooling seqtab and taxa would result in highly similar ASVs being merged as well as exact matches. 

Am I misunderstanding how this process works? My MiSeq and HiSeq runs have processed fine but some have not merged or differ in length. I am using paired-ends but my understanding is that the exact ASV sequence in the rowname may differ slightly between runs for sequences that are similar and would have been assigned the same ASV if they were processed together. 

Again perhaps I am misunderstanding how this works. 

For example, my ITS data - the MiSeq run has 3304 ASVs, the HiSeq 19138 ASVs, (about 100 vs 700 samples on each respectively). Of the MiSeq seqtab colnames (the ASVs) 940 are not exact matches to the respective list from the HiSeq run.

The combined file has 20078 ASVs, 940 more than the HiSeq alone. 

The samples on the MiSeq run are a random subset drawn from a single experimental study so we are expecting te same signal to come through.

**Comment by benjjneb** [comment]
>I've merged my fungal ASVs but found that this has only combined the two seqtab tables in such a way that exactly matching ASVs are pooled but non-exact matches by rowname are not. I thought that pooling seqtab and taxa would result in highly similar ASVs being merged as well as exact matches.
>
>Am I misunderstanding how this process works?

I think so. Merging will only "merge" exactly identical sequences, so the behavior you are describing is expected.

> My MiSeq and HiSeq runs have processed fine but some have not merged or differ in length. I am using paired-ends but my understanding is that the exact ASV sequence in the rowname may differ slightly between runs for sequences that are similar and would have been assigned the same ASV if they were processed together.

If you have the same forward/reverse primers, and if your truncation parameters were the same in both runs (and you didn't use `truncQ`) then no artefactual length variation should have been introduced, and sequences should merge (or not) based on exact matching.

> For example, my ITS data - the MiSeq run has 3304 ASVs, the HiSeq 19138 ASVs, (about 100 vs 700 samples on each respectively). Of the MiSeq seqtab colnames (the ASVs) 940 are not exact matches to the respective list from the HiSeq run.

This is not a surprising number by itself. It is normal that there are unique ASVs to any subset of samples, not for bioinformatic reasons, but because of both sample processing artefacts (e.g. batch-specific contaminants) and rare ASVs that will be present in a few samples that might be found in one batch and not the other.

To determine if there is any cause for concern, I would ask how many reads those MiSeq-specific ASVs represent out of the entire MiSeq dataset. If it is a small fraction, then this is relatively normal and not obviously a cause for concern.

**Comment by drinkermoth** [comment]
They represent 50,950 reads out of 3,001,315 total reads on the run, or 1.70% so not a lot I suppose!

 Of the non-zero values in the 940 ASVs that are not shared, they have the following distribution:   Min: 2.00    1st Qu:8.00   Median :17.00  Mean: 44.81  3rd Qu.: 40.00 Max. :1126.00  

so I suspect many of them will pass QC and enter the analysis stage as detected "real" values. 
![novelITSASVs](https://user-images.githubusercontent.com/7210494/90522513-b735f500-e163-11ea-8863-c1c4afc170cc.png)

Any advice as to filtering or processing these ASVs. Equally any warnings or suggestions about combining HiSeq and MiSeq data or am I on the right track?

**Comment by benjjneb** [comment]
I think you are doing AOK, and <2% of reads in ASVs specific to the MiSeq part of the study is not a cause for concern. It's actually a rather low number -- it would make me feel better that I was getting similar measurements from the MiSeq and HiSeq runs.

**Comment by drinkermoth** [comment]
Thanks!

**Comment by drinkermoth** [comment]
I have reopened this as, looking at the 16S data I have more of a problem. 49.9% (1,691,244/3,385,982) of the reads in the MiSeq run lack exact matches in the HiSeq run. The runs share 16131 ASVs and 27627 of the ASVs in the MiSeq run are absent from the HiSeq run. 

HiSeq filtAndTrim:
> out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, maxN = 0, maxEE = c(2, 2), truncLen=c(220,200), 
+                      truncQ = 2, minLen = 50, rm.phix = TRUE, compress = TRUE,
+                      multithread = TRUE)  # on windows, set multithread = FALSE

MiSeq filtAndTrim is identical:
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, maxN = 0, maxEE = c(2, 2), truncLen=c(220,200), 
                     truncQ = 2, minLen = 50, rm.phix = TRUE, compress = TRUE,
                     multithread = TRUE)  # on windows, set multithread = FALSE

They are not clearly differentiated by size:
![seqLenMissing density](https://user-images.githubusercontent.com/7210494/90651394-dea8c280-e234-11ea-8a95-0fe77ae79fc6.png)
![seqLenShared density](https://user-images.githubusercontent.com/7210494/90651397-df415900-e234-11ea-9b11-ee8e80d8cd03.png)

They don't appear to vary by samples which might indicate a primer tag bias, except for positive samples which tend to have only exact matches. 

So all in all I'm a little confused about how best to mode forward. For some reason the HiSeq run proved very difficult to analyse through DADA2 (3days and 13h with 325.87Gb RAM, 30 cores multithreaded, the assign taxa step was the longest). However I can run the MiSeq run through in 6h or so. So if I can make changes to the MiSeq run that will bring it into line with the HiSeq run that would be a quick fix. 

Do you have any ideas as to what might be causing the issue?

**Comment by benjjneb** [comment]
Two things I would recommend:

(1) I'm confused about the difference in your previous post and this one on the fraction or MiSeq reads in ASVs not in the Hiseq run (am I wrong, but before it was <2% and now its ~50%?). But, this can be attempted to be reduced in two ways. First would be to give the Miseq calls information about the ASVs that were found int he Hiseq runs. That would be done by first defining a set of AVSs you feel good about from the Hiseq runs (call them `seqs.hiseq`), perhaps because they each occur in at least 2 (or 3 or whatever) Hiseq samples. Then, pass those to the denoising of the Miseq runs as so: `dada(..., priors=seqs.hiseq)'. This will seed the denoising with information about what was seen in the hiseq runs. Secondarily, running `collapseNoMismatch` on the combined Miseq/Hiseq table could possibly help, but given the questionable computational performance of that function at the moment I would have that be a secondary option.

(2) `assignTaxonomy` has always been a bit of a laggard in our package in terms of attention to its performance. We've made some recent improvements. I would recommend installing the latest dev version of dada2 from the Github repository for future large-scale `assignTaxonomy` jobs. It should be significantly faster, and this may solve the vector error you posted above.

**Comment by benjjneb** [comment]
> I gained some ASVs but no change in the proportion of MiSeq ASVs missing from the HiSeq run. Was this the right way to set it up?

Yes that looks good to me.

> I'm going to try to merge the runs in the traditional way, then run collapseNoMismatch( ) on that with a decent number of cores and a couple of hundred Gb of RAM, see if that improves anything.

It should work but perhaps slowly. Note that `collapseNoMismatch` is single-core, so you'll get no gain from more cores (we have an issue open to potentially address that: https://github.com/benjjneb/dada2/issues/626. The issue there is really one of optimizing the underlying algorithm for the era of growing dataset sizes.

> My MiSeq processing was 3-4 times faster now but with a slight increase in max RAM usage, perhaps from including the priors.

Yes that is expected, using priors will yeild a slight RAM usage increase.

> I got the same error trying to assign taxonomy to the combined seqtab, but I do have the seqtab... it got up to around 7-8Gb of RAM, way within alloaction.

Hmm. That I don't understand then. I recommend trying to upgrade to the current devel version 1.17.3 (or later), but if that doesn't help this might be a seperate issue that needs investigation.

**Comment by drinkermoth** [comment]
Digging into this some more I have identified that in my 16S data there are no ASVs that occur in more than 30 samples of the 700 in the HISeq run and none that occur in more than 2 samples in the MiSeq run! I think that the ASV generation algorithms have been too sensitive. 

Do you have any idea how I might reduce this sensitivity, apart from running collapseNoMismatch( ) as before suggested? 

This is probably why my 16S sequencing runs were taking so long to process through the dada2 pipeline before and hopefulyl will fix my issues merging these runs.

**Comment by benjjneb** [comment]
> Digging into this some more I have identified that in my 16S data there are no ASVs that occur in more than 30 samples of the 700 in the HISeq run and none that occur in more than 2 samples in the MiSeq run! I think that the ASV generation algorithms have been too sensitive.
> This is probably why my 16S sequencing runs were taking so long to process through the dada2 pipeline before and hopefulyl will fix my issues merging these runs.

Yeah there is something going on here that we don't understand yet.

One possibility: Did your library prep introduce any "artefactual" length variation through the use of techniques like "heterogeneity spacers" or variable length primers or variable length sample barcodes? That kind of library prep is not handled appropriately by standard DADA2 trimming, and could result in the kind of issues seen here: slower processing, and poor overlap between ASVs across samples.

**Comment by drinkermoth** [comment]
> like "heterogeneity spacers" 

Yes... For some reason, I had got it into my head that leaving these in would be no problem. Obviously I removed them with CUTADAPT in the ITS1 samples. In this one, I had previously removed them using cutadapt but then got very poor output with many samples failing to pass filtering steps. I felt that this might be due to cutadapt incorrectly removing files with heterogeneity spacers and removed cutadapt from the workflow, I realise now that that was incorrect. 

It appears my issue was due to read length and I suspect what had been happening was that I was including too much of the read in samples with primers removed, which caused them to be filtered out for low quality. When primers were included, reads were able to be merged despite their shorter length and were of higher quality so were not filtered out, but then varied in length in each sample. 

Now I have employed cutadapt in the workflow again but shortened the read length using the _truncLen_ argument by about 20-30bp. I'm finding that I no longer have the problem of reads not passing filtering anymore, and also only 0.77% of reads are from ASVs in the miseq run that are absent from the HiSeq run. A lot better than 49% from previous runs. The processing time also went down to about 16h with 30 cores and 162.5G of RAM for half a HiSeq 16S run. Hopefully, the runs will merge and assign taxa fine. 

This is a good example of occasions where problems can be misdiagnosed as one issue, due to more fundamental problems with the bioinformatic setup. Thank's Ben for guiding me through this one, I'm sorry for taking up so much of your time on a basic error!
